### PR TITLE
[core] Manifest and Statistic support query by timestamp

### PR DESCRIPTION
**Pull request by herefree**
### Purpose

fllow up #4291
Linked issue: close #xxx

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java [hunk 1]
```diff
@@ -200,6 +200,8 @@ private static List<ManifestFileMeta> allManifests(FileStoreTable dataTable) {
         SnapshotManager snapshotManager = dataTable.snapshotManager();
         Long snapshotId = coreOptions.scanSnapshotId();
         String tagName = coreOptions.scanTagName();
+        Long timestampMills = coreOptions.scanTimestampMills();
+
         Snapshot snapshot = null;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java:205]: Variable 'snapshot' initializer 'null' is redundant.
Change "Snapshot snapshot = null;" to "Snapshot snapshot;".

File: paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java [hunk 2]
```diff
@@ -212,12 +214,16 @@ private static List<ManifestFileMeta> allManifests(FileStoreTable dataTable) {
...
+                        String.format("Specified scan.tag-name %s is not exist.", tagName));
             }
+            snapshot = dataTable.tagManager().tag(tagName).trimToSnapshot();
+        } else if (timestampMills != null) {
+            snapshot = snapshotManager.earlierOrEqualTimeMills(timestampMills);
+        } else {
+            snapshot = snapshotManager.latestSnapshot();
         }
 
         if (snapshot == null) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java:229]: Add a warn log?

File: paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java [hunk 3]
```diff
@@ -212,12 +214,16 @@ private static List<ManifestFileMeta> allManifests(FileStoreTable dataTable) {
...
             }
             snapshot = snapshotManager.snapshot(snapshotId);
-        } else {
-            if (!StringUtils.isEmpty(tagName) && dataTable.tagManager().tagExists(tagName)) {
-                snapshot = dataTable.tagManager().tag(tagName).trimToSnapshot();
-            } else {
-                snapshot = snapshotManager.latestSnapshot();
+        } else if (!StringUtils.isEmpty(tagName)) {
+            if (!dataTable.tagManager().tagExists(tagName)) {
+                throw new RuntimeException(
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java:219]: throw new RuntimeException(
                        String.format("Specified parameter %s = %s is not exist.", SCAN_TAG_NAME.key(), tagName));

File: paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java [hunk 4]
```diff
@@ -212,12 +214,16 @@ private static List<ManifestFileMeta> allManifests(FileStoreTable dataTable) {
                                 snapshotId, earliestSnapshotId, latestSnapshotId));
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/ManifestsTable.java:214]: String.format(
                                "Specified parameter %s = %s is not exist, you can set it in range from %s to %s.",
                                SCAN_SNAPSHOT_ID.key(), snapshotId, earliestSnapshotId, latestSnapshotId));

File: paimon-core/src/test/java/org/apache/paimon/table/system/ManifestsTableTest.java [hunk 5]
```diff
@@ -135,6 +135,24 @@ public void testReadManifestsFromSpecifiedTagName() throws Exception {
         assertThat(result).containsExactlyElementsOf(expectedRow);
     }
 
+    @Test
+    public void testReadManifestsFromSpecifiedTimestampMillis() throws Exception {
+        Thread.sleep(10);
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/system/ManifestsTableTest.java:140]: Why sleep?

File: paimon-core/src/test/java/org/apache/paimon/table/system/ManifestsTableTest.java [hunk 6]
```diff
@@ -135,6 +135,24 @@ public void testReadManifestsFromSpecifiedTagName() throws Exception {
...
+        write(table, GenericRow.of(3, 1, 1), GenericRow.of(3, 2, 1));
+        long time = System.currentTimeMillis();
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/system/ManifestsTableTest.java:142]: inline "long time = System.currentTimeMillis();"

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala [hunk 7]
```diff
@@ -83,13 +83,16 @@ abstract class AnalyzeTableTestBase extends PaimonSparkTestBase {
     spark.sql(s"INSERT INTO T VALUES ('1', 'a', 1, 1)")
     spark.sql(s"INSERT INTO T VALUES ('2', 'aaa', 1, 2)")
     Assertions.assertEquals(0, spark.sql("select * from `T$statistics`").count())
-
+    Thread.sleep(10)
     spark.sql(s"ANALYZE TABLE T COMPUTE STATISTICS")
+    val time1 = System.currentTimeMillis
```
Comment (wwj6591812) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala:88]: inline "val time1 = System.currentTimeMillis"

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala [hunk 8]
```diff
@@ -83,13 +83,16 @@ abstract class AnalyzeTableTestBase extends PaimonSparkTestBase {
...
 
     spark.sql(s"INSERT INTO T VALUES ('3', 'b', 2, 1)")
     spark.sql(s"INSERT INTO T VALUES ('4', 'bbb', 3, 2)")
 
+    Thread.sleep(10)
     spark.sql(s"ANALYZE TABLE T COMPUTE STATISTICS")
+    val time2 = System.currentTimeMillis
```
Comment (wwj6591812) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala:95]: inline "val time2 = System.currentTimeMillis"

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala [hunk 9]
```diff
@@ -118,6 +121,12 @@ abstract class AnalyzeTableTestBase extends PaimonSparkTestBase {
         Row(2, 0, 2, "{ }"))
     }
 
+    withSQLConf("spark.paimon.scan.timestamp-millis" -> time1.toString) {
+      checkAnswer(
+        sql("SELECT snapshot_id, schema_id, mergedRecordCount, colstat FROM `T$statistics`"),
```
Comment (wwj6591812) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/AnalyzeTableTestBase.scala:126]: I think you can modify "colstat" to "colStat" in StatisticTable. And then modify other relevant codes.

**Comment by herefree** [comment]
@wwj6591812 thanks for your review,I have completed the modification.

**Comment by wwj6591812** [comment]
+1

File: paimon-core/src/main/java/org/apache/paimon/table/system/StatisticTable.java [hunk 10]
```diff
@@ -72,7 +72,7 @@ public class StatisticTable implements ReadonlyTable {
                             new DataField(1, "schema_id", new BigIntType(false)),
                             new DataField(2, "mergedRecordCount", new BigIntType(true)),
                             new DataField(3, "mergedRecordSize", new BigIntType(true)),
-                            new DataField(4, "colstat", SerializationUtils.newStringType(true))));
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/system/StatisticTable.java:75]: please don't modify this, it is better to keep compability.

Comment (herefree) [reply paimon-core/src/main/java/org/apache/paimon/table/system/StatisticTable.java]: addressed

**Comment by JingsongLi** [comment]
+1

**Pull request merged**